Remove AutoDeleteSmallVector and AutoFreeSmallVector

michicc opened the pull request:
Remove the last remains of SmallVector. For the replacement, `std::unique_ptr` takes care of the memory management. Use of references and move semantics makes ownership clear and reduces potentials for memory leaks. Current compilers should be able to optimize most of the copy/moves out, resulting in very little runtime overhead.

This got a bit bigger than expected and picked up some unrelated fixes I saw when in the area.

Commits 41b6916, 0266dcb, and b27d0af could be split out if needed.

michicc commented:
There's a slight work-around for clang-3.8 included (extra `std::move` that shouldn't be needed). Starting with clang-3.9 (verified with godbolt) this wouldn't be needed anymore.

michicc commented:
Now with more `std::string` to please @nielsmh 😁 

Also, why the heck is GitHub displaying the commits in a mostly random order?

M3Henry commented:
Oh good, this was on my TODO list

LordAro commented on src/widgets/dropdown.cpp:
```diff
@@ -51,11 +51,11 @@ void DropDownListStringItem::Draw(int left, int right, int top, int bottom, bool
  * @warning All items in the list need to be derivates of DropDownListStringItem.
  */
-/* static */ int DropDownListStringItem::NatSortFunc(const DropDownListItem * const *first, const DropDownListItem * const * second)
+/* static */ int DropDownListStringItem::NatSortFunc(std::unique_ptr<const DropDownListItem> const &first, std::unique_ptr<const DropDownListItem> const &second)
```
might make more sense to pass in the raw pointers here?

michicc replied:
It's not about making sense, it's about what `std::sort` expects. And in this usage, a raw pointer won't work.

LordAro commented on src/gfx_layout.cpp:
```diff
@@ -795,21 +808,21 @@ Point Layouter::GetCharPosition(const char *ch) const
 const char *Layouter::GetCharAtPosition(int x) const
 {
-	const ParagraphLayouter::Line *line = this->front();
+	auto const &line = this->front();
```
const auto (+ others)

michicc replied:
Will do.

LordAro commented on src/game/game_text.hpp:
```diff
@@ -29,12 +29,12 @@ struct LanguageStrings {
-	AutoDeleteSmallVector<LanguageStrings *> compiled_strings; ///< The compiled strings per language, first must be English/the master language!.
-	StringList string_names;                                   ///< The names of the compiled strings.
+	std::vector<std::unique_ptr<LanguageStrings>> raw_strings;      ///< The raw strings per language, first must be English/the master language!.
+	std::vector<std::shared_ptr<LanguageStrings>> compiled_strings; ///< The compiled strings per language, first must be English/the master language!.
```
how difficult would it be to "fix" the use of shared_ptr here?
or, perhaps the various places that "own" the pointer could be documented?

michicc replied:
Very easily, it is only `GameStrings::cur_language` that points to one of the compiled string entries. We store a raw pointer now and just assume it's never going to be invalidated; still possible.

In pure though, raw pointers are bad and should never be used. Of course, the codebase is far, far, far from that point now. So, philosophical question time: do the "right" thing, or the pragmatic thing?

LordAro reviewed (commented):
use of noexcept is interesting. I've seen varying opinions on how worthwhile it actually is...

michicc committed:
Fix: Don't crash if reading a GS string file from disk produces an error.

The raw_strings vector may not include NULLs as no consumer can deal with it.

michicc committed:
Codechange: Use override specifier for text layout classes.

michicc committed:
Fix: Forgotten override keywords for DropDownListIconItem.

michicc committed:
Codechange: Store text layout runs directly as values in a std::vector instead of heap allocated.

This reduces memory allocations and heap fragmentation.

michicc committed:
Codechange: Replace AutoDeleteSmallVector with direct std::vector use in text layout code.

michicc committed:
Codechange: Use std::vector instead of AutoDeleteSmallVector in GS text handling.

michicc committed:
Codechange: Switch DropDownList to directly use std::vector, thus making AutoDeleteSmallVector obsolete.

DropDownListItem are strongly managed using std::unique_ptr to ensure leak-free handling. Appropriate use
of move-semantics make intent a lot clearer than parameter comments and allows the compiler to generate
copy-free code for most situations.

michicc committed:
Codechange: If something is a vector of strings, use a vector of strings instead of an AutoFreeSmallVector.

michicc committed:
Remove: AutoFreeSmallVector.

The last use was for storing a list of memory blocks. As the way these lists are accessed is very
specific, it is easier to just write an explicit destructor instead of trying to exactly match the behaviour.

michicc commented:
Rebased to drop the already applied OSX commit.

PeterN commented:
Does the change from returning a `DropDownList *` to returning a `DropDownList` mean it makes a copy?

michicc commented:
In the literal sense: No. `std::vector` is a moveable type, which means the return statement will resolve to a move assignment and not a copy assignment. Moving a vector is quite cheap.

In the less literal sense: Many current compilers will perform copy elision where possible, which means even the move assignment is likely to be optimized out.
